### PR TITLE
Reduce `FSM<TState, TData>` allocations

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

Reduced some unnecessary allocations inside the `FSM<TState, TData>` class.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [ ] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Changes in public API reviewed, if any.

### Latest `v1.4` Benchmarks 

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.2006 (21H2)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.201
  [Host]     : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT
  DefaultJob : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT


```
|           Method | MsgCount |     Mean |    Error |   StdDev | Ratio | RatioSD |      Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|----------------- |--------- |---------:|---------:|---------:|------:|--------:|-----------:|----------:|----------:|----------:|
|     BenchmarkFsm |  1000000 | 772.8 ms | 13.20 ms | 12.35 ms |  1.74 |    0.06 | 68000.0000 | 4000.0000 | 1000.0000 |    287 MB |
| BenchmarkUntyped |  1000000 | 444.3 ms |  8.88 ms |  9.50 ms |  1.00 |    0.00 | 12000.0000 | 4000.0000 | 1000.0000 |     56 MB |

### This PR's Benchmarks

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.2006 (21H2)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.201
  [Host]     : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT
  DefaultJob : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT


```
|           Method | MsgCount |     Mean |    Error |   StdDev | Ratio | RatioSD |      Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|----------------- |--------- |---------:|---------:|---------:|------:|--------:|-----------:|----------:|----------:|----------:|
|     BenchmarkFsm |  1000000 | 662.5 ms | 12.99 ms | 13.34 ms |  1.47 |    0.05 | 35000.0000 | 4000.0000 | 1000.0000 |    150 MB |
| BenchmarkUntyped |  1000000 | 453.5 ms |  8.97 ms | 11.35 ms |  1.00 |    0.00 | 13000.0000 | 3000.0000 | 1000.0000 |     56 MB |

48% memory reduction, 14% throughput improvement

### CODE REVIEW

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.verified.txt [hunk 1]
```diff
@@ -736,7 +736,7 @@ namespace Akka.Actor
             public override int GetHashCode() { }
             public override string ToString() { }
         }
-        public sealed class Event<TD> : Akka.Actor.INoSerializationVerificationNeeded
+        public struct Event<TD> : Akka.Actor.INoSerializationVerificationNeeded
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.verified.txt:739]: Technically a binary breaking change maybe, but it's source compatible.

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.verified.txt [hunk 2]
```diff
@@ -773,10 +773,10 @@ namespace Akka.Actor
         {
             public static Akka.Actor.FSMBase.StateTimeout Instance { get; }
         }
-        public class State<TS, TD> : System.IEquatable<Akka.Actor.FSMBase.State<TS, TD>>
+        public sealed class State<TS, TD> : System.IEquatable<Akka.Actor.FSMBase.State<TS, TD>>
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.verified.txt:776]: `sealed` the State class for better perf.

File: src/core/Akka/Actor/FSM.cs [hunk 3]
```diff
@@ -539,15 +540,16 @@ public State(TS stateName, TD stateData, TimeSpan? timeout = null, Reason stopRe
             /// </summary>
             internal State<TS, TD> WithNotification(bool notifies)
             {
+                // don't bother allocating even a stack type if the notifies value is identical.
+                if (Notifies == notifies)
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/FSM.cs:544]: Don't return a new copy of this object if the current `Notifies`  value matches the new one (it usually does.)

File: src/core/Akka/Actor/FSM.cs [hunk 4]
```diff
@@ -750,7 +752,7 @@ public void StartWith(TState stateName, TData stateData, TimeSpan? timeout = nul
         /// <returns>Descriptor for staying in the current state.</returns>
         public State<TState, TData> Stay()
         {
-            return GoTo(_currentState.StateName).WithNotification(false);
+            return _currentState.WithNotification(false);
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/FSM.cs:755]: Biggest memory savings - don't allocate a new `State` object each time we call `Stay`. Re-use the same value.

**Review by Aaronontheweb — COMMENTED** [review]
Described API and memory allocation changes.

File: src/core/Akka/Actor/FSM.cs (hunk 4 above)
Comment (Aaronontheweb) [reply src/core/Akka/Actor/FSM.cs:755]: FWIW, there's no benefit to copying the same state over and over again here. The `Reason`, `Replies` properties et al don't have any effects when used here in combination with `WithNotification(false)`.

**Comment by Aaronontheweb** [comment]
Looked at making `State<TData, TState>` into a `struct` as well but that's a significant breaking API change (we have lots of FSM behaviors that return `null` in order to trigger `Unhandled` behavior) that doesn't yield much benefit once we stop returning new states during `Stay()` operations.

**Comment by Aaronontheweb** [comment]
Looks like this change definitely caused an issue with the `FSMTimingSpec`. Going to need to look at that before I ask for review.

**Comment by Aaronontheweb** [comment]
> Looks like this change definitely caused an issue with the `FSMTimingSpec`. Going to need to look at that before I ask for review.

Spec could also be racy - worked on my machine.

**Comment by Aaronontheweb** [comment]
> > Looks like this change definitely caused an issue with the `FSMTimingSpec`. Going to need to look at that before I ask for review.
> 
> Spec could also be racy - worked on my machine.

Nope, definitely an error related to my changes. Will investigate.

**Comment by Aaronontheweb** [comment]
Benchmarks based on latest changes.

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.2006 (21H2)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.201
  [Host]     : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT
  DefaultJob : .NET 6.0.3 (6.0.322.12309), X64 RyuJIT


```
|           Method | MsgCount |     Mean |    Error |   StdDev | Ratio | RatioSD |      Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|----------------- |--------- |---------:|---------:|---------:|------:|--------:|-----------:|----------:|----------:|----------:|
|     BenchmarkFsm |  1000000 | 641.1 ms | 12.73 ms | 14.15 ms |  1.35 |    0.05 | 35000.0000 | 4000.0000 | 1000.0000 |    150 MB |
| BenchmarkUntyped |  1000000 | 474.1 ms |  9.38 ms | 10.04 ms |  1.00 |    0.00 | 12000.0000 | 4000.0000 | 1000.0000 |     56 MB |

**Review by Arkatufus — APPROVED** [review]
LGTM

**Pull request merged**